src/util/XDRStream.h

```diff
@@ -118,10 +136,78 @@ class XDRInputFileStream
         {
             throw xdr::xdr_runtime_error("malformed XDR file");
         }
+
         xdr::xdr_get g(mBuf.data(), mBuf.data() + sz);
         xdr::xdr_argpack_archive(g, out);
         return true;
     }
+
+    template <typename T>
+    bool
+    readPage(T& out, LedgerKey const& key, size_t pageSize)
+    {
+        ZoneScoped;
+        if (mBuf.size() < pageSize)
+        {
+            mBuf.resize(pageSize);
+        }
+
+        if (!mIn.read(mBuf.data(), pageSize))
+        {
+            if (mIn.eof())
+            {
+                mIn.clear(std::ios_base::eofbit);
+            }
+            else
+            {
+                return false;
```

Comment:
I also _think_ this `failbit` handling is wrong (as is the handling in `readOne` above, which of course I wrote) though I've only been banging my head against this part of `iostreams` for 25 years and it's possible I'm getting it wrong again. I _think_ if `mIn` is false after the read then either there was an IO fault (`badbit`) or the operation logically failed (`failbit`) and if EOF was the _cause_ of the logical failure then `eofbit` will also be set. So we want to treat anything _other_ than `eofbit` as an actual problem and escalate by throwing, not returning false. I think.

Reply:
(I will fiddle with this locally to try to confirm)

Reply:
(have posted changes to this PR to resolve this)